third_party/amd/lib/TritonAMDGPUToLLVM/ElementwiseOpToLLVM.cpp

```diff
@@ -1330,6 +1330,93 @@ struct Exp2OpConversion
   bool ftz;
 };
 
+struct LogOpConversion
+    : ElementwiseOpConversionBase<mlir::math::LogOp, LogOpConversion> {
+  using ElementwiseOpConversionBase<
+      mlir::math::LogOp, LogOpConversion>::ElementwiseOpConversionBase;
+
+  explicit LogOpConversion(LLVMTypeConverter &typeConverter,
+                           ModuleAxisInfoAnalysis &axisInfoAnalysis, bool ftz,
+                           PatternBenefit benefit)
+      : ElementwiseOpConversionBase(typeConverter, axisInfoAnalysis, benefit),
+        ftz(ftz) {}
+
+  SmallVector<Value> createDestOps(mlir::math::LogOp op, OpAdaptor adaptor,
+                                   ConversionPatternRewriter &rewriter,
+                                   Type elemTy, MultipleOperandsRange operands,
+                                   Location loc) const {
+    // For non-FP32 input, call __ocml_log_f64 for higher-precision calculation
+    if (elemTy.getIntOrFloatBitWidth() != 32)
+      return {};
+
+    // On AMD backend, both intrinsics are lowered to v_log_f32 instruction,
+    // which flushes input and output denorms. `llvm.amdgcn.log.f32` is directly
+    // lowered to v_log_f32, while `llvm.log.f32` has some auxiliary processes,
+    // including denorm preservation, handling inf, and change the base of log2
+    // to natural log.
+    StringRef funcName = ftz ? "llvm.amdgcn.log.f32" : "llvm.log.f32";
+    Type funcType = getFunctionType(elemTy, operands[0]);
+    LLVM::LLVMFuncOp funcOp =
+        appendOrGetExternFuncOp(rewriter, op, funcName, funcType);
+
+    Value intrinsicsOutput =
+        LLVM::createLLVMCallOp(rewriter, loc, funcOp, operands[0]).getResult();
+
+    // In case of non-ftz, `llvm.log.f32` has already changed the base, so we
+    // can safely return here.
+    if (!ftz)
+      return {intrinsicsOutput};
+
+    // In case of ftz, `llvm.amdgcn.log.f32` is lowered directly to v_log_f32,
+    // which is a log2 operator. Thus we need to adjust the result to simulate
+    // natural log.
+    const double loge2 = 0.693147180559945286;
+    return {fmul(f32_ty, intrinsicsOutput, f32_val(loge2))};
```

Comment:
It's worth mentioning that there are some auxiliary instructions in the assembly code to deal with infinity:

```llvm
define float @v_log_f32(float %in) {
; GFX900-GISEL-LABEL: v_log_f32:
; GFX900-GISEL:       ; %bb.0:
; GFX900-GISEL-NEXT:    s_waitcnt vmcnt(0) expcnt(0) lgkmcnt(0)
; GFX900-GISEL-NEXT:    v_mov_b32_e32 v1, 0x800000
; GFX900-GISEL-NEXT:    v_mov_b32_e32 v2, 0x4f800000
; GFX900-GISEL-NEXT:    v_cmp_lt_f32_e32 vcc, v0, v1
; GFX900-GISEL-NEXT:    v_cndmask_b32_e32 v1, 1.0, v2, vcc
; GFX900-GISEL-NEXT:    v_mul_f32_e32 v0, v0, v1
; GFX900-GISEL-NEXT:    v_log_f32_e32 v0, v0
; GFX900-GISEL-NEXT:    v_mov_b32_e32 v1, 0x3f317217
; GFX900-GISEL-NEXT:    v_mov_b32_e32 v3, 0x3377d1cf
; GFX900-GISEL-NEXT:    v_mul_f32_e32 v2, 0x3f317217, v0
; GFX900-GISEL-NEXT:    v_fma_f32 v1, v0, v1, -v2
; GFX900-GISEL-NEXT:    v_fma_f32 v1, v0, v3, v1
; GFX900-GISEL-NEXT:    v_add_f32_e32 v1, v2, v1
; GFX900-GISEL-NEXT:    v_mov_b32_e32 v2, 0x7f800000
; GFX900-GISEL-NEXT:    v_cmp_lt_f32_e64 s[4:5], |v0|, v2
; GFX900-GISEL-NEXT:    v_cndmask_b32_e64 v0, v0, v1, s[4:5]
; GFX900-GISEL-NEXT:    v_mov_b32_e32 v1, 0x41b17218
; GFX900-GISEL-NEXT:    v_cndmask_b32_e32 v1, 0, v1, vcc
; GFX900-GISEL-NEXT:    v_sub_f32_e32 v0, v0, v1
; GFX900-GISEL-NEXT:    s_setpc_b64 s[30:31]
  %result = call float @llvm.log.f32(float %in)
  ret float %result
}
```

Do we need to replay all of these in the conversion?

Reply:
These are handled by the backend codegen, right?

Reply:
What is the behavior on the nv side? Do they not handle denorm and infinity if going down the "fast" path?

Reply:
On the nv side, it handles inf no matter ftz or non-ftz. For denorm, it chooses between `@llvm.nvvm.fma.rn.ftz.f` or `@llvm.nvvm.fma.rn.f` according to `__CUDA_FTZ`.